Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp

```diff
@@ -67,13 +69,48 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
         currentRootLayer->setTransform(matrix);
 
     bool sceneHasRunningAnimations = currentRootLayer->applyAnimationsRecursively(MonotonicTime::now());
-    m_textureMapper->beginPainting(flipY ? TextureMapper::FlipY::Yes : TextureMapper::FlipY::No);
-    m_textureMapper->beginClip(TransformationMatrix(), FloatRoundedRect(clipRect));
 
+    WTFBeginSignpost(this, CollectDamage);
+    currentRootLayer->collectDamage(*m_textureMapper);
+    WTFEndSignpost(this, CollectDamage);
+    WebCore::Damage boundsDamage;
+    const auto& frameDamage = ([this, &boundsDamage]() -> const WebCore::Damage& {
+        const auto& damage = lastDamage();
+        if (m_propagateDamage != Damage::ShouldPropagate::No && !damage.isInvalid()) {
+            if (m_settings.unifyDamagedRegions()) {
+                boundsDamage.add(damage.bounds());
+                return boundsDamage;
+            }
+            return damage;
+        }
+        return WebCore::Damage::invalid();
+    })();
+
+    FloatRoundedRect actualClipRect(clipRect);
+    if (m_client && m_settings.useDamagingInformationInCompositing()) {
+        // FIXME: Decide what to do with valid-yet-empty damage (bounds:0,0,0,0).
+        m_client->damageRenderTargets(frameDamage);
+
+        const Damage& damageSinceLastRenderTargetUse = m_client->renderTargetDamage();
+        if (!damageSinceLastRenderTargetUse.isInvalid())
+            actualClipRect = (FloatRoundedRect)damageSinceLastRenderTargetUse.bounds();
+    }
+
+    WTFBeginSignpost(this, PaintTextureMapperLayerTree);
+    m_textureMapper->beginPainting(flipY ? TextureMapper::FlipY::Yes : TextureMapper::FlipY::No);
+    m_textureMapper->beginClip(TransformationMatrix(), actualClipRect);
     currentRootLayer->paint(*m_textureMapper);
-    m_fpsCounter.updateFPSAndDisplay(*m_textureMapper, clipRect.location(), matrix);
     m_textureMapper->endClip();
     m_textureMapper->endPainting();
+    WTFEndSignpost(this, PaintTextureMapperLayerTree);
+
+    if (m_fpsCounter.isActive()) {
+        m_textureMapper->beginPainting(flipY ? TextureMapper::FlipY::Yes : TextureMapper::FlipY::No);
+        m_textureMapper->beginClip(TransformationMatrix(), FloatRoundedRect(clipRect));
+        m_fpsCounter.updateFPSAndDisplay(*m_textureMapper, clipRect.location(), matrix);
+        m_textureMapper->endClip();
+        m_textureMapper->endPainting();
+    }
```

Comment:
Why do we need to separate the fps counter painting here now?

Reply:
FPS counter is not taking part in damaging the frame, so if damage propagation is enabled, now it's possible that the clipping region will be limited to the area that does not include FPS counter and therefore counter won't be actually painted (it will be painted outside the clipping region).

Technically - as the FPS counter is not working with skia atm. - this change currently prevents regression only in cairo builds.